crates/solver/src/settlement_ranker.rs

```diff
@@ -0,0 +1,167 @@
+}
+
+impl SettlementRanker {
+    /// Discards settlements without user orders and settlements which violate price checks.
+    /// Logs info and updates metrics about the out come of this run loop for each solver.
+    fn discard_illegal_settlements(
+        &self,
+        solver: &Arc<dyn Solver>,
+        settlements: Result<Vec<Settlement>, SolverRunError>,
+        external_prices: &ExternalPrices,
+    ) -> Vec<Settlement> {
+        let name = solver.name();
+        match settlements {
+            Ok(mut settlement) => {
+                for settlement in &settlement {
+                    tracing::debug!(solver_name = %name, ?settlement, "found solution");
+                }
+
+                // Do not continue with settlements that are empty or only liquidity orders.
+                let settlement_count = settlement.len();
+                settlement.retain(solver_settlements::has_user_order);
+                if settlement_count != settlement.len() {
+                    tracing::debug!(
+                        solver_name = %name,
+                        "settlement(s) filtered containing only liquidity orders",
+                    );
+                }
+
+                if let Some(max_settlement_price_deviation) = &self.max_settlement_price_deviation {
+                    let settlement_count = settlement.len();
+                    settlement.retain(|settlement| {
+                        settlement.satisfies_price_checks(
+                            solver.name(),
+                            external_prices,
+                            max_settlement_price_deviation,
+                            &self.token_list_restriction_for_price_checks,
+                        )
+                    });
+                    if settlement_count != settlement.len() {
+                        tracing::debug!(
+                            solver_name = %name,
+                            "settlement(s) filtered for violating maximum external price deviation",
+                        );
+                    }
+                }
+
+                let outcome = match settlement.is_empty() {
+                    true => SolverRunOutcome::Empty,
+                    false => SolverRunOutcome::Success,
+                };
+                self.metrics.solver_run(outcome, name);
+
+                settlement
+            }
+            Err(err) => {
+                let outcome = match err {
+                    SolverRunError::Timeout => SolverRunOutcome::Timeout,
+                    SolverRunError::Solving(_) => SolverRunOutcome::Failure,
+                };
+                self.metrics.solver_run(outcome, name);
+                tracing::warn!(solver_name = %name, ?err, "solver error");
+                vec![]
+            }
+        }
+    }
+
+    /// Appends a merged settlement if multiple simple settlements canlegally be merged into a
+    /// single more efficient settlement.
+    fn append_merged_settlement(
```

Comment:
One refactor that I think would be useful here, would be to push settlement merging to the `SingleOrderSolver` wrapper type (i.e. the type that turns an `impl SingleOrderSolver` into an `impl Solver`).

The reason I suggest this is that, eventually, with the solver+driver co-location, each solver+driver will present a single settlement (so would have to have already shuffled and merged single order trades and simulated and ranked them to pick the best one).

Reply:
Great suggestion!
Making the merge part of the settlement ranking didn't feel right anyway.

Reply:
This also allows our `trait Solver` to return a single settlement instead of a `Vec<Settlement>` and push merging logic only to single order solvers.

Reply:
Refactoring the merge into the `SingleOrderSolver` would blow up this PR. I will do that refactor first and then rebase the `SettlemenRanker` on top of that.

>This also allows our trait Solver to return a single settlement instead of a Vec<Settlement> and push merging logic only to single order solvers.

I'm a bit hesitant about this one. If we only return the merged settlement a single flaky single-order-settlement could result in the solver not competing (if it causes the merged settlement to revert for some reason).

Reply:
> flaky single-order-settlement could result in the solver not competing

Shouldn't it be doing its simulations for the merging? I assumed that it would simulate the various proposed merged settlements and pick the best one from those.

Reply:
>  could result in the solver not competing

We should still do it. Currently single order solvers gain an unfair advantage compared to other solvers because the driver does the merging for them. It should be something for solvers to decide how big they want to make their settlements.

Reply:
> We should still do it.

So, I took a look at the other PR and agree that we don't necessarily need to change the `trait Solver` to return a single `Settlement` for now.

Once we finish co-locating Driver+Solver, we will automatically get this (since each Driver+Solver combo will only get a single settlement to propose). I think we can wait to refactor some of these abstractions as we finish the colocation logic instead of trying to do it all now.

That being said, I do think we should still do it at some point. The `merge_settlements` will be a no-op for all Driver+HTTPSolver instances we have, so it seems like logic that is specific to `SingleOrderSolver` implementations, and should make its way there eventually.